### PR TITLE
Allow to register para using relay proof

### PR DESCRIPTION
**Pull request by tmpolaczyk**
WIP

### CODE REVIEW

File: pallets/registrar/src/lib.rs [hunk 1]
```diff
@@ -585,6 +540,152 @@ pub mod pallet {
...
+            origin: OriginFor<T>,
+            para_id: ParaId,
+            parathread_params: Option<ParathreadParamsTy>,
+            relay_proof_block_number: u32,
+            relay_storage_proof: sp_trie::StorageProof,
+            manager_signature: cumulus_primitives_core::relay_chain::Signature,
+            genesis_data: ContainerChainGenesisData<T::MaxLengthTokenSymbol>,
+        ) -> DispatchResult {
+            let account = ensure_signed(origin)?;
+            if !T::AllowRegisterWithRelayProof::get() {
```
Comment (girazoki) [inline pallets/registrar/src/lib.rs:558]: Maybe we can prevent this at the origin level? (EnsureNever)

File: pallets/registrar/src/lib.rs [hunk 2]
```diff
@@ -585,6 +540,152 @@ pub mod pallet {
...
+            // Should include:
+            // * para_id, in case the manager has more than 1 para in the relay
+            // * accountid in tanssi, to ensure that the creator role is assigned to the desired account
+            // * relay_proof_block_number or something time based, to make the signature expiring
+            //   ^ this can make UX very bad, but maybe we can check a conservative range and not the
+            //   exact block number? eg by using (relay_proof_block_number / 128) as the signature msg
+            // * parathread_params, to avoid registering a parathread as parachain and viceversa
+            //   ^ maybe not needed, the relay manager should trust the tanssi creator
+            // * also don't need genesis_data as part of the signature
+            let signature_msg: Vec<u8> = (para_id, &account, relay_proof_block_number).encode();
```
Comment (girazoki) [inline pallets/registrar/src/lib.rs:599]: We should indeed speak about this yes, as it is important, A few ideas:

- the signature includes something network-specific (so that a signature cannot be re-used across networks). Maybe the relay block hash against which we are veryfing could be a good sign
- By default by signing the block hash this is expiring, as this is one of our stored last N block hashes

File: test/scripts/build-spec.sh [hunk 3]
```diff
@@ -7,7 +7,7 @@ set -e
 cd $(dirname $0)/..
 
 mkdir -p specs
-../target/release/container-chain-simple-node build-spec --disable-default-bootnode --add-bootnode "/ip4/127.0.0.1/tcp/33049/ws/p2p/12D3KooWHVMhQDHBpj9vQmssgyfspYecgV6e3hH1dQVDUkUbCYC9" --parachain-id 2000 --raw > specs/template-container-2000.json
-../target/release/container-chain-frontier-node build-spec --disable-default-bootnode --add-bootnode "/ip4/127.0.0.1/tcp/33050/ws/p2p/12D3KooWFGaw1rxB6MSuN3ucuBm7hMq5pBFJbEoqTyth4cG483Cc" --parachain-id 2001 --raw > specs/template-container-2001.json
+../target/release/container-chain-simple-node build-spec --disable-default-bootnode --parachain-id 2000 --raw > specs/template-container-2000.json
+../target/release/container-chain-frontier-node build-spec --disable-default-bootnode --parachain-id 2001 --raw > specs/template-container-2001.json
```
Comment (tmpolaczyk) [inline test/scripts/build-spec.sh:11]: Undo, it turns out those bootnodes are not needed, but this shouldn't be in this PR

**Review by ParthDesai — APPROVED** [review]

File: pallets/registrar/src/benchmark_blob.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            // Account generated by
+            // create_funded_user::<T>("caller", 0, T::DepositAmount::get());
+            let account32 = AccountId32::decode(
+                &mut [
+                    92u8, 119, 3, 63, 206, 138, 144, 69, 130, 74, 102, 144, 187, 249, 156, 109,
+                    178, 105, 80, 47, 10, 141, 29, 42, 0, 133, 66, 213, 105, 10, 7, 73,
+                ]
+                .as_slice(),
+            )
+            .unwrap();
```
Comment (girazoki) [inline pallets/registrar/src/benchmark_blob.rs:144]: An idea to make these run faster:remove the y parameteres in both register and deregister. In the `register` benchmarks we can completely ignore the y parameter as we have a mapping now for the chains pending verification.

In the `deregister` extrinsic, let's assume x and y is always max, since it's what we are doing anyways in the pallet. That way we iterate only over z

File: pallets/registrar/src/lib.rs [hunk 5]
```diff
@@ -1182,3 +1343,22 @@ where
...
 }
+
+// TODO: import this from dancekit
+pub const REGISTRAR_PARAS_INDEX: &[u8] =
+    &hex_literal::hex!["3fba98689ebed1138735e0e7a5a790abcd710b30bd2eab0352ddcc26417aa194"];
+
+// Need to copy ParaInfo from
+// polkadot-sdk/polkadot/runtime/common/src/paras_registrar/mod.rs
+// Because its fields are not public...
+// TODO: import this from dancekit
```
Comment (girazoki) [inline pallets/registrar/src/lib.rs:1354]: Let's make a PR to parity to make them public, it does not make any sense that they are not

File: pallets/registrar/src/lib.rs [hunk 6]
```diff
@@ -743,6 +819,91 @@ pub mod pallet {
...
+
+            T::RegistrarHooks::para_marked_valid_for_collating(para_id);
+
+            Ok(())
+        }
+
+        /// Relay parachain manager signature message. Includes:
+        /// * para_id, in case the manager has more than 1 para in the relay
+        /// * accountid in tanssi, to ensure that the creator role is assigned to the desired account
+        /// * relay_storage_root, to make the signature network-specific, and also make it expire
```
Comment (girazoki) [inline pallets/registrar/src/lib.rs:897]: Nice!

File: pallets/registrar/src/lib.rs [hunk 7]
```diff
@@ -620,6 +586,116 @@ pub mod pallet {
...
+                    cumulus_primitives_core::relay_chain::AccountId,
+                    cumulus_primitives_core::relay_chain::Balance,
+                >>(key.as_slice())
+                .map_err(|_| Error::<T>::InvalidRelayStorageProof)?;
+            if relay_para_info.is_some() {
+                return Err(Error::<T>::ParaStillExistsInRelay.into());
+            }
+
+            // Take the deposit immediately and give it to origin account
+            if let Some(asset_info) = RegistrarDeposit::<T>::take(para_id) {
```
Comment (girazoki) [inline pallets/registrar/src/lib.rs:682]: This needs to come after the deregister. Because:

- if the deregister has already been made by the manager, then it is very likely that the deregistration in the relay has not finished yet. Therefore this slashing condition will apply. And the deposit will still have not been unreserved because we do it when the deregistration completes.

- In contrast if we do it after the deregister, the deregister should error if you are trying to deregister something that is already in the deregistering process.

What do you think? If agreed, can we create a test for this?

File: test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+                const credits2001 = (await polkadotJs.query.servicesPayment.blockProductionCredits(2001)).toJSON();
+                expect(credits2001).to.be.null;
+            },
+        });
+    },
+});
+
+// Creating a relay storage proof in dev tests is not possible, because there is no relay chain, but we can re-use the
+// proof from setValidationData, which includes the registrar->paras entries because we added them.
+async function fetchStorageProofFromValidationData(polkadotJs) {
```
Comment (girazoki) [inline test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts:148]: love this

File: pallets/registrar/src/lib.rs (hunk 7 above)
Comment (girazoki) [reply pallets/registrar/src/lib.rs:682]: Nevermind, I am still with the past mentality in which extrinsics where not transactional. Forget about it :)

Comment (tmpolaczyk) [reply pallets/registrar/src/lib.rs:682]: The deposit needs to be taken before, because there is an edge case in deregister that allows the deposit to be removed immediately, so it will not exist if we try to take the deposit after the call to deregister. But I should add tests.

File: test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts (hunk 8 above)
Comment (tmpolaczyk) [reply test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts:148]: I hate it

File: test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+                // TODO: fix once we have types
+                expect(parasScheduled.toJSON()).to.deep.equal([2000]);
+
+                // Checking that in session 2 paras are registered
+                await jumpSessions(context, 2);
+
+                // Expect now paraIds to be registered
+                const parasRegistered = await polkadotJs.query.registrar.registeredParaIds();
+                // TODO: fix once we have types
+                expect(parasRegistered.toJSON()).to.deep.equal([2000]);
```
Comment (girazoki) [inline test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts:82]: Can we check we get the deposit?

Comment (girazoki) [reply test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts:82]: that we slash the deposit I mean

File: test/suites/dev-tanssi/register-with-relay-proof/test_register_with_relay_proof.ts [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+            polkadotJs = context.polkadotJs();
+
+            // We must generate the same account as in service.rs, using get_ed25519_pairs
+            // Instead of having to implement that function in javascript, we can just hardcode the private key here
+            // Create a new keyring because we need to use ed25519
+            const relayKeyring = new Keyring({
+                type: "ed25519",
+            });
+            relayManager = relayKeyring.addFromUri(
+                "0x3132333435363738393031323334353637383930313233343536373839303132"
```
Comment (girazoki) [inline test/suites/dev-tanssi/register-with-relay-proof/test_register_with_relay_proof.ts:27]: let's put this into a constnat

File: test/suites/dev-tanssi/register-with-relay-proof/test_register_with_relay_proof.ts [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                const givenFreeCredits2 = await polkadotJs.query.servicesPayment.givenFreeCredits(paraId);
+                expect(givenFreeCredits2.isNone).to.be.true;
+            },
+        });
+    },
+});
+
+// Creating a relay storage proof in dev tests is not possible, because there is no relay chain, but we can re-use the
+// proof from setValidationData, which includes the registrar->paras entries because we added them.
+async function fetchStorageProofFromValidationData(polkadotJs) {
```
Comment (girazoki) [inline test/suites/dev-tanssi/register-with-relay-proof/test_register_with_relay_proof.ts:170]: Move to utils

**Review by girazoki — APPROVED** [review]
Besides small comments (the benchmark can be addressed in a follow up PR), looks good @tmpolaczyk, good job!

File: test/suites/dev-tanssi/register-with-relay-proof/test_register_with_relay_proof.ts (hunk 10 above)
Comment (tmpolaczyk) [reply test/suites/dev-tanssi/register-with-relay-proof/test_register_with_relay_proof.ts:27]: In utils or here?

File: test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts (hunk 9 above)
Comment (tmpolaczyk) [reply test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts:82]: In this test I cannot do it because that parachain is registered in genesis, so it doesn't have a deposit.

Comment (girazoki) [reply test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts:82]: mmm it's a pity

Comment (girazoki) [reply test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts:82]: can we register another chain insted?

Comment (tmpolaczyk) [reply test/suites/dev-tanssi/register-with-relay-proof/test_deregister_with_relay_proof.ts]: I registered chain 2003 instead
